modules/javafx.base/src/main/java/javafx/beans/value/ObservableValue.java

```diff
@@ -251,4 +252,63 @@ default ObservableValue<T> orElse(T constant) {
     }
+
+    /**
+     * Returns an {@code ObservableValue} that holds this value whenever the given
+     * condition evaluates to {@code true}, otherwise holds the last value when
+     * {@code condition} became {@code false}. The value is updated whenever this
+     * {@code ObservableValue} changes, unless the condition currently evaluates
+     * to {@code false}.
+     * <p>
+     * The returned {@code ObservableValue} only observes this value when the given
+     * {@code condition} evaluates to {@code true}. This allows this {@code ObservableValue}
+     * and the conditional {@code ObservableValue} to be garbage collected if neither is
+     * otherwise strongly referenced when {@code condition} becomes {@code false}.
+     * <p>
+     * A currently observed binding will observe its source, which means it will not be eligible
+     * for garbage collection while source isn't. However, using {@code when} this {@code ObservableValue}
+     * can still be eligible for garbage collection when the condition is {@code false} and the
+     * conditional itself is also eligible for garbage collection.
+     * <p>
+     * Returning {@code null} from the given condition is treated the same as
+     * returning {@code false}.
+     * <p>
+     * For example:
+     * <pre>{@code
+     * ObservableValue<Boolean> condition = new SimpleBooleanProperty(true);
+     * ObservableValue<String> globalProperty = new SimpleStringProperty("A");
+     * ObservableValue<String> whenProperty = property.when(isShowing);
+     *
+     * // observe whenProperty, which will in turn observe globalProperty
+     * whenProperty.addChangeListener((ov, old, current) -> System.out.println(current));
+     *
+     * globalProperty.setValue("B");  // "B" is printed
+     *
+     * condition.setValue(false);
+     *
+     * // After condition becomes false, whenProperty stops observing globalProperty; condition
+     * // and whenProperty may now be eligible for GC despite being observed by the ChangeListener
+     *
+     * globalProperty.setValue("C");  // nothing is printed
+     * globalProperty.setValue("D");  // nothing is printed
+     *
+     * condition.setValue(true);  // globalProperty is observed again, and "D" is printed
+     * }</pre>
+     * Another example:
+     * <pre>{@code
+     * Label label = ... ;
+     * ObservableValue<String> globalProperty = new SimpleStringProperty("A");
+     *
+     * // bind label's text to a global property only when it is showing:
+     * label.textProperty().bind(globalProperty.when(label::isShowingProperty));
+     * }</pre>
+     * @param condition a boolean {@code ObservableValue}, cannot be {@code null}
+     * @return an {@code ObservableValue} that holds this value whenever the given
+     *     condition evaluates to {@code true}, otherwise holds the last seen value;
+     *     never returns {@code null}
+     * @since 20
+     */
+    default ObservableValue<T> when(ObservableValue<Boolean> condition) {
```

Comment:
Discussion: name? `when`, `conditionOn`, other suggestions?  Note that this is not the same as a possible future method `filter`, as a `filter` would still keep observing its source at all times (and would use a `Predicate` as parameter).

Reply:
`when` sounds fine, but we need to be careful about its relation to the `When` class.

This reminds me also that some of the new functionality in `ObservableValue` can be used as a replacement for `When`, maybe we should note that like we did with `Bindings.select`.

Reply:
Do you want me to make this part of the PR? Basically, `Bindings.when` with `map` or `flatMap` depending on whether or not you refer to a value or another property in `then` or `otherwise`.

Reply:
Changing the documentation for `When` should not be part of this PR I think. What we should do maybe is add a note that this method should not be confused with `Bindings.when`. I'll let others weigh in.

Reply:
I like `conditionOn`.

Reply:
I actually like `asLongAs` that was mentioned in another comment, but it looks odd :)

Lately, as I was reading some unrelated codebases and I came up with `updateWhile` (or `updateWhen`). The reason is that names like `when` and `conditionOn`, and even the restricted `if` and `while`, tell me just half the story. They tell me what happens "when", but not what happens "when not". The `bindings.When` binding is phrased as "`when` condition `then` A, `otherwise` B", which is I think it great. This binding is phrased as "`when` condition", but then what and what otherwise?

Because the binding updates its value when/while/as long as the condition holds and *doesn't update when the condition doesn't hold* (it just stays the same), I think that the phrasing "`update while` condition" (and don't update while not condition) gives a better description of the functionality.

Reply:
Does `asLongAs` imply recurrent nature?

Reply:
I don't see why not.